File: miner.py
import sys
import time
import requests
import threading

# --- QTP MINER ---
# A simple command line miner for the QTP network
# Run this to start mining and earning QTP rewards

NODE_URL     = "http://localhost:5001"
POLL_INTERVAL = 5  # seconds between mining attempts

def print_header():
    print("=" * 55)
    print("  QUBIT TOPCOIN (QTP) MINER")
    print("  Quantum-Resistant. For Everyone. Forever.")
    print("=" * 55)
    print("")

def get_node_status():
    try:
        response = requests.get(f"{NODE_URL}/status", timeout=5)
        return response.json()
    except Exception:
        return None

def mine_block():
    try:
        response = requests.get(f"{NODE_URL}/mine", timeout=300)
        return response.json()
    except Exception as e:
        return {"error": str(e)}

def check_balance(address):
    try:
        response = requests.get(
            f"{NODE_URL}/balance/{address}",
            timeout=5
        )
        return response.json().get("balance", 0)
    except Exception:
        return None

def mining_loop(miner_address):
    print(f"Mining to address: {miner_address}")
    print(f"Connected to node: {NODE_URL}")
    print(f"Press Ctrl+C to stop mining\n")
    print("-" * 55)

    blocks_mined  = 0
    total_earned  = 0.0
    start_time    = time.time()

    while True:
        try:
            # Check node status
            status = get_node_status()
            if not status:
                print(f"[{timestamp()}] Node unreachable. "
                      f"Retrying in {POLL_INTERVAL}s...")
                time.sleep(POLL_INTERVAL)
                continue

            pending = status.get("pending_txs", 0)
            blocks  = status.get("blocks", 0)

            if pending == 0:
                print(f"[{timestamp()}] No pending transactions. "
                      f"Waiting {POLL_INTERVAL}s... "
                      f"(chain height: {blocks})")
                time.sleep(POLL_INTERVAL)
                continue

            # Mine pending transactions
            print(f"[{timestamp()}] {pending} pending transaction(s) found. "
                  f"Mining...")

            result = mine_block()

            if "error" in result:
                print(f"[{timestamp()}] Mining error: {result['error']}")
                time.sleep(POLL_INTERVAL)
                continue

            if "message" in result and "mined" in result["message"]:
                blocks_mined += 1
                reward        = result.get("reward", 50.0)
                total_earned += reward
                elapsed       = time.time() - start_time

                print(f"[{timestamp()}] ✓ Block mined!")
                print(f"  Hash:         {result.get('hash', '')[:40]}...")
                print(f"  Reward:       {reward} QTP")
                print(f"  Total earned: {total_earned} QTP "
                      f"({blocks_mined} blocks)")
                print(f"  Mining time:  {elapsed:.0f}s total")

                # Check wallet balance
                balance = check_balance(miner_address)
                if balance is not None:
                    print(f"  Wallet balance: {balance} QTP")
                print("-" * 55)

            time.sleep(POLL_INTERVAL)

        except KeyboardInterrupt:
            print(f"\n\n{'=' * 55}")
            print(f"  MINING STOPPED")
            print(f"{'=' * 55}")
            print(f"  Blocks mined:  {blocks_mined}")
            print(f"  Total earned:  {total_earned} QTP")
            print(f"  Time elapsed:  {time.time() - start_time:.0f}s")
            print(f"{'=' * 55}\n")
            sys.exit(0)

def timestamp():
    return time.strftime("%H:%M:%S")

def main():
    print_header()

    # Get miner address
    if len(sys.argv) > 1:
        miner_address = sys.argv[1]
    else:
        print("Usage: python3 miner.py <your_qtp_address>")
        print("Example: python3 miner.py GbtE3vPeomLSRNN2k9DsDfy8YL164JXMw2")
        print("")
        miner_address = input("Enter your QTP wallet address: ").strip()
        if not miner_address:
            print("No address provided. Exiting.")
            sys.exit(1)

    print("")

    # Check node is reachable before starting
    print("Connecting to node...")
    status = get_node_status()
    if not status:
        print(f"Cannot reach node at {NODE_URL}")
        print("Make sure your node is running with: python3 node.py 5001")
        sys.exit(1)

    print(f"Connected. Chain height: {status.get('blocks', 0)} blocks")
    print("")

    # Start mining
    mining_loop(miner_address)

if __name__ == "__main__":
    main()

File: qtpcore/src/bin/miner.rs

fn register_miner(node_url: &str, address: &str) -> bool {
    let url    = format!("{}/set_miner/{}", node_url, address);
    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .unwrap();
    match client.post(&url).send() {
        Ok(r)  => r.status().is_success(),
        Err(_) => false,
    }
}

// =========================================
//   QTP — Qubit TopCoin CPU Miner
//   Cori Testnet
//   Quantum-Resistant. For Everyone. Forever.
// =========================================

use std::time::{Duration, Instant};
use std::thread;

const NODE_URL:       &str = "http://localhost:5003";
const POLL_INTERVAL:  u64  = 5; // seconds between checks

fn print_header() {
    println!("");
    println!("=========================================");
    println!("  QTP CPU MINER");
    println!("  Quantum-Resistant. For Everyone. Forever.");
    println!("  Network: Cori Testnet");
    println!("=========================================");
    println!("");
}

fn get_status(node_url: &str) -> Option<serde_json::Value> {
    let url    = format!("{}/status", node_url);
    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
        .unwrap();
    match client.get(&url).send() {
        Ok(r)  => r.json().ok(),
        Err(_) => None,
    }
}

fn mine_block(node_url: &str, _miner_address: &str) -> Option<serde_json::Value> {
    let url    = format!("{}/mine", node_url);
    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(300)) // 5 min timeout for mining
        .build()
        .unwrap();
    match client.get(&url).send() {
        Ok(r)  => r.json().ok(),
        Err(_) => None,
    }
}

fn get_balance(node_url: &str, address: &str) -> u64 {
    let url = format!("{}/balance/{}", node_url, address);
    match reqwest::blocking::get(&url) {
        Ok(r) => {
            if let Ok(data) = r.json::<serde_json::Value>() {
                data["balance"].as_u64().unwrap_or(0)
            } else {
                0
            }
        }
        Err(_) => 0,
    }
}

fn timestamp() -> String {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();

    let hours   = (now % 86400) / 3600;
    let minutes = (now % 3600) / 60;
    let seconds = now % 60;
    format!("{:02}:{:02}:{:02}", hours, minutes, seconds)
}

fn main() {
    let args: Vec<String> = std::env::args().collect();

    // Get miner address from args or prompt
    let miner_address = if args.len() > 1 {
        args[1].clone()
    } else {
        println!("Usage: miner <your_qtp_address> [node_url]");
        println!("Example: miner GRgnXA7g5fz58EU58nriyuUq4FwkxhL3vQ");
        println!("");
        print!("Enter your QTP wallet address: ");
        use std::io::Write;
        std::io::stdout().flush().unwrap();
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        input.trim().to_string()
    };

    if miner_address.is_empty() {
        println!("No address provided. Exiting.");
        return;
    }

    let node_url = if args.len() > 2 {
        args[2].clone()
    } else {
        NODE_URL.to_string()
    };

    print_header();
    println!("Mining to:  {}", miner_address);
    println!("Node:       {}", node_url);
    println!("Press Ctrl+C to stop");
    println!("-----------------------------------------");
    println!("");

    // Check node is reachable
    print!("Connecting to node...");
    match get_status(&node_url) {
        Some(status) => {
            println!(" connected");
            println!(
                "Chain height: {} blocks | Difficulty: {}",
                status["blocks"].as_u64().unwrap_or(0),
                status["difficulty"].as_u64().unwrap_or(0)
            );
        }
        None => {
            println!(" FAILED");
            println!("Cannot reach node at {}", node_url);
            println!("Make sure your node is running.");
            return;
        }
    }

    // Register miner address with node
    print!("Registering miner address...");
    if register_miner(&node_url, &miner_address) {
        println!(" done");
    } else {
        println!(" failed (node may not support this yet)");
    }

    println!("");

    // Mining loop
    let mut blocks_mined:  u64   = 0;
    let mut total_earned:  u64   = 0;
    let start_time               = Instant::now();

    loop {
        // Check for pending transactions
        match get_status(&node_url) {
            Some(status) => {
                let pending = status["pending_txs"].as_u64().unwrap_or(0);
                let height  = status["blocks"].as_u64().unwrap_or(0);

                if pending == 0 {
                    println!(
                        "[{}] Waiting for transactions... (height: {})",
                        timestamp(), height
                    );
                    thread::sleep(Duration::from_secs(POLL_INTERVAL));
                    continue;
                }

                println!(
                    "[{}] {} pending transaction(s) found. Mining...",
                    timestamp(), pending
                );

                // Mine the block
                let mine_start = Instant::now();
                match mine_block(&node_url, &miner_address) {
                    Some(result) => {
                        if result.get("hash").is_some() {
                            blocks_mined += 1;

                            let reward = result["reward"].as_u64()
                                .unwrap_or(5_000_000_000);
                            total_earned += reward;

                            let elapsed    = mine_start.elapsed();
                            let total_time = start_time.elapsed();

                            println!("[{}] ✓ Block mined!", timestamp());
                            println!(
                                "  Hash:         {}",
                                result["hash"].as_str().unwrap_or("unknown")
                            );
                            println!(
                                "  Miner:        {}",
                                &miner_address[..20]
                            );
                            println!(
                                "  Time:         {:.1}s",
                                elapsed.as_secs_f64()
                            );
                            println!(
                                "  Reward:       {:.8} QTP",
                                reward as f64 / 100_000_000.0
                            );
                            println!(
                                "  Blocks mined: {}",
                                blocks_mined
                            );
                            println!(
                                "  Total earned: {:.8} QTP",
                                total_earned as f64 / 100_000_000.0
                            );
                            println!(
                                "  Mining time:  {:.0}s total",
                                total_time.as_secs_f64()
                            );

                            // Check actual wallet balance
                            let balance = get_balance(&node_url, &miner_address);
                            println!(
                                "  Wallet:       {:.8} QTP",
                                balance as f64 / 100_000_000.0
                            );
                            println!("-----------------------------------------");
                        } else {
                            println!(
                                "[{}] {}",
                                timestamp(),
                                result["message"].as_str().unwrap_or("No transactions")
                            );
                        }
                    }
                    None => {
                        println!(
                            "[{}] Mining error — node unreachable",
                            timestamp()
                        );
                    }
                }
            }
            None => {
                println!(
                    "[{}] Node offline. Retrying in {}s...",
                    timestamp(), POLL_INTERVAL
                );
            }
        }

        thread::sleep(Duration::from_secs(POLL_INTERVAL));
    }
}
